### ISSUE TITLE
Is it possible to a create a websocket server in crow that transmit continuously to a webbrowser (client)

### ISSUE THREAD

**Issue by akjt** [CrowCpp/Crow#515]
I have now reached were we want to create a websocket server which streams data continuously based on internal event. For example, a certain internal event happened and we want to transmit some data to the client. 

Currently, with the tutorials/docs provided I only see the option of having a server that transmits data on request by the client which is the "onmessage" function. Is that try? Does crow not support what I am looking for ?

**Comment by The-EDev** [comment]
A websocket handler (Route) is designed to respond to client events, in a similar fashion to an HTTP route, unlike the latter however, the [websocket connection](https://crowcpp.org/master/reference/classcrow_1_1websocket_1_1_connection.html) object (which can be obtained from the `onopen` handler at the earliest) has a number of methods to send messages to the client as long as the connection is established and regardless of whether or not the client made any request immediately prior.

**Comment by akjt** [comment]
Thanks very much. 
Can you refer me to where to look at it when it comes to the number of methods that can send messages without prior request?
For example, I assume that onmessage is not the one to use, as to me this sounds something triggered on an event made by a client

**Comment by The-EDev** [comment]
https://crowcpp.org/master/reference/classcrow_1_1websocket_1_1_connection.html

any public method that starts with `send_`, the 2 main ones are `send_text` and `send_binary`. as long as you have a reference to `crow::websocket::connection`, you can use these methods.

Ultimately the client would still need to contact the server to create the websocket connection itself, this is when `onaccept` and `onopen` will run, `onaccept` provides you the opportunity to prevent the websocket connection from being created based on the HTTP request. While `onopen` provides the `crow::websocket::connection` reference once the connection has been successfully established.

I hope this clarifies matters :)

**Comment by The-EDev** [comment]
closing the issue as resolved, @akjt let me know if you're still facing difficulties and I'll reopen it.

**Comment by akjt** [comment]
That was really helpful!

**Comment by akjt** [comment]
Thanks very much.
Can you refer me to where to look at it when it comes to the number of
methods that can send messages without prior request?
For example, I assume that onmessage is not the one to use, as to me this
sounds something triggered on an event made by a client


søn. 7. aug. 2022 kl. 17.10 skrev Farook Al-Sammarraie <
***@***.***>:

**Comment by The-EDev** [comment]
@akjt as long as you have the connection object, you can use `send_text` or `send_binary` (there are also `send_ping` and `send_pong`) to send data without a prompt from the client

**Comment by jake-arkinstall** [comment]
> I assume that onmessage is not the one to use

In the [websocket example code](https://github.com/CrowCpp/Crow/blob/master/examples/websocket/example_ws.cpp), onmessage is only used as a trigger (from one connection) to flush a message to all connections. The clue that it isn't a request/response interaction is that the connection from the `onmessage` call is discarded. You don't need to do sending from inside onmessage - you can use the same approach (loop through client connections, invoke `send_text` or `send_binary`) outside of the crow handling logic. I'll show that in an example below.

But first, take a look at the example and see what's happening. Clients get registered upon making a connection (they get added to the `users` set). When *any* of the clients sends a message to the Crow server, it gets distributed out to *all* users, like a simple chat server. Note that, while from the *sender*'s perspective it's a request/response behaviour (they sent the message, they get it back), from the perspective of *other* users the message appears without making a request. Open it in two tabs and you'll see what I mean.

This example demonstrates that messages can be sent ad-hoc, using `connection->send_text` or `connection->send_binary`, but unfortunately it demonstrates two concepts at the same time (receiving a message and sending a message), and that might be a source of confusion. I think additional examples might help - @The-EDev I'll be happy to provide some, though I first encountered websockets this morning for my own purposes, so I'll take some more time to learn them properly.

Now let's turn to an example of using it outside of the event handler. For completeness, let's have two scenarios: If a user sends a message, it gets sent to all other users. Every 1 second, an additional message is sent from the server. The routing logic is identical to the example, so I'll focus on what needs to change to achieve it.

The example currently runs on the main thread:

```cpp
app.port(40080)
      .multithreaded()
      .run();
```

so to send a message from another thread, we can simply do:

```cpp
std::atomic<bool> do_continue = true; // flip to false to stop sending automated messages
std::thread send_thread([&](){
    while(do_continue) {
        std::this_thread::sleep_for(std::chrono::seconds(1)); // wait 
        std::lock_guard<std::mutex> _(mtx);                   // lock (don't clash with other messages)
        for (auto u : users){                                 // iterate through users
              u->send_text("Automated message");              // send
        }
    }
});
app.port(40080).multithreaded().run();      // start the crow server
do_continue = false;                        // when crow shuts down, tell the other thread to close
send_thread.join();                         // join the other thread
```

In other words, as long as your "sending" code has access to the list of connections (or `users`) and the sending mutex (if additional communication is happening to), you can easily send to all connections.
